### PR TITLE
add stable option to argsort op

### PR DESCRIPTION
**Pull request by haoyouab**
This PR adds stable sort option and let users choose stable/unstable sort while using `edsl::argsort`. Notes are listed below.

1. No new algorithm is added but just modification of the original selection sort. Reference can be found in [wikipedia](https://en.wikipedia.org/wiki/Selection_sort). In short:
> Selection sort can be implemented as a stable sort if, rather than swapping in step 2, the minimum value is inserted into the first position and the intervening values shifted up. However, this modification either requires a data structure that supports efficient insertions or deletions, such as a linked list, or it leads to performing Θ(n2) writes.

Since we don't use linked list in this case, we should be aware that this modification can incur performance impact.

2. Stable sort is needed in Ops like `TopK` when comparing with OV reference implementation. [tf.argsort](https://www.tensorflow.org/api_docs/python/tf/argsort) has boolean `stable` parameter, too.

3. For now the argsort algorithm is working correctly but its time complexity is O(n^2). In the long term, we should implement an O(n*log(n)) argsort algorithm.

### CODE REVIEW

File: plaidml/edsl/__init__.py [hunk 1]
```diff
@@ -774,19 +774,30 @@ class SortDirection(enum.IntEnum):
...
+
+    UNSTABLE = 1
+    """Sort a tensor in an unstable way."""
+
+
+def argsort(x, axis=-1, direction=SortDirection.ASC, stability=SortStability.UNSTABLE):
     """Returns the indices of a tensor ``x`` that give its sorted order along
     an ``axis``, where -1 represents the last axis, in either ascending or
-    descending ``direction``.
+    descending ``direction`` and with either stable or unstalbe ``stability``.
```
Comment (tzerrell) [inline plaidml/edsl/__init__.py:790]: typo: `unstalbe` -> `unstable`

File: plaidml/edsl/__init__.py [hunk 2]
```diff
@@ -774,19 +774,30 @@ class SortDirection(enum.IntEnum):
...
 
     Args:
         x (Tensor): The tensor to be sorted.
         axis (TensorDim): The axis to perform sorting on.
         direction (SortDirection): The direction of value comparison.
+        stability (SortStability): The stability of value comparison.
```
Comment (tzerrell) [inline plaidml/edsl/__init__.py:796]: Maybe say "Whether the sort is stable" instead? Stability is more about whether certain reorderings are legal than about how comparisons are made.

File: plaidml/edsl/__init__.py [hunk 3]
```diff
@@ -774,19 +774,30 @@ class SortDirection(enum.IntEnum):
     """Sort a tensor in descending order."""
 
 
-def argsort(x, axis=-1, direction=SortDirection.ASC):
+class SortStability(enum.IntEnum):
+    """Stability for a sort operation."""
+
+    STABLE = 0
+    """Sort a tensor in a stable way."""
```
Comment (tzerrell) [inline plaidml/edsl/__init__.py:781]: For clarity, perhaps instead give a definition of a stable sort, e.g. "Preserve the initial order of equivalent values".

File: plaidml/edsl/__init__.py [hunk 4]
```diff
@@ -774,19 +774,30 @@ class SortDirection(enum.IntEnum):
...
+
+    UNSTABLE = 1
+    """Sort a tensor in an unstable way."""
```
Comment (tzerrell) [inline plaidml/edsl/__init__.py:784]: Similarly, could define, e.g. "Permit reordering of equivalent values".

File: plaidml/edsl/tests/edsl_test.cc [hunk 5]
```diff
@@ -2243,6 +2243,70 @@ TEST_F(CppEdsl, ArgSort1d) {
...
+  //    5: 56.79,  6: 57.52,  7: 7.11,   8: 39.48,  9:  5.9
+  //   10: 14.81, 11: 66.23, 12: 20.25, 13: 66.05, 14: 64.5
+  //   15: 71.07, 16: 67.6,  17: 54.42, 18: 87.59, 19: 80.02
+  // sorted:
+  //    9:  5.9,   7:  7.11, 10: 14.81, 12: 20.25,  2: 27.74
+  //    8: 39.48,  3: 43.69, 17: 54.42,  4: 55.79,  5: 56.79
+  //    6: 57.52, 14: 64.5,  13: 66.05, 11: 66.23, 16: 67.6
+  //   15: 71.07, 19: 80.02,  1: 81.69,  0: 81.69, 18: 87.59
+  std::vector<int32_t> output = {
+      9, 7, 10, 12, 2, 8, 3, 17, 4, 5, 6, 14, 13, 11, 16, 15, 19, 1, 0, 18  //
```
Comment (tzerrell) [inline plaidml/edsl/tests/edsl_test.cc:2298]: Hmm, do we actually want to be testing that an unstable sort will _necessarily_ reorder equivalent values?

That said I'm not certain of the best way to test an unstable sort... perhaps use Slice to remove the last few (unstable) values?

File: plaidml/edsl/tests/edsl_test.py [hunk 6]
```diff
@@ -666,6 +666,106 @@ def test_argsort_1d(self):
...
+            5,
+            6,
+            14,
+            13,
+            11,
+            16,
+            15,
+            19,
+            1,
+            0,
```
Comment (tzerrell) [inline plaidml/edsl/tests/edsl_test.py:764]: See my earlier comment on testing that enforces reordering in an unstable sort.

**Review by tzerrell — COMMENTED** [review]
Looks generally good to me although I don't have a good sense of what sorting algorithms are reasonable for our use cases. @mars1026 does the stable algorithm look reasonable to you?

Beyond that, I have a comment on testing, and some minor documentation suggestions.

File: plaidml/edsl/tests/edsl_test.cc (hunk 5 above)
Comment (haoyouab) [reply plaidml/edsl/tests/edsl_test.cc:2298]: Or I can just remove the 'unstable' tests.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 7]
```diff
@@ -1015,42 +1015,41 @@ struct ArgSortOpConversion : public OpConversionPattern<tile::ArgSortOp> {
       minVal = rewriter.create<LoadOp>(loc, minValVar, zeroIndex).getResult();
       // Is compVal smaller than minVal? If so, update the allocs
       Value orderPred;
-      if (elementType.isSignedInteger()) {
```
Comment (mars1026) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: Why are these cases being reordered?

File: pmlc/dialect/tile/ir/ops.td [hunk 8]
```diff
@@ -97,7 +97,8 @@ def ArgSortOp : MaterializeOp<"argsort"> {
   let arguments = (ins
     RankedTensorOf<[AnyScalar]>:$tensor,
     IndexAttr:$axis,
-    SortDirection:$direction);
+    SortDirection:$direction,
+    SortStability:$stability);
   let results = (outs RankedTensorOf<[SI32]>:$result);
   let assemblyFormat = [{
     $direction $tensor `[` $axis `]` attr-dict `:` functional-type(operands, $result)
```
Comment (flaub) [inline pmlc/dialect/tile/ir/ops.td:104]: We should adjust this custom format so that it includes the stability.

**Comment by flaub** [comment]
My basic feeling is that we shouldn't change the argsort interface unless absolutely necessary. In this sense, I feel like we should only support stable sorting. It seems like we can always write a better algorithm later, or even make a parallel version, but that should all be implementation details behind the argsort interface.

**Comment by mars1026** [comment]
On Feb 23, 2021, at 2:03 PM, Frank Laub <notifications@github.com<mailto:notifications@github.com>> wrote:



My basic feeling is that we shouldn't change the argsort interface unless absolutely necessary. In this sense, I feel like we should only support stable sorting. It seems like we can always write a better algorithm later, or even make a parallel version, but that should all be implementation details behind the argsort interface.

The hitch is that stability has a cost; it limits our choice of algorithms and not toward the better-performing options. Requiring the client to request stability when they need it therefore preserves preserves implementation flexibility.  This is the approach TensorFlow has taken with their argsort implementation: the 'stable' parameter defaults to False, and you have to specify stable=True if you want that guarantee - even though they have yet to provide a non-stable sort implementation.

File: plaidml/edsl/edsl.h [hunk 9]
```diff
@@ -805,8 +811,10 @@ enum class SortDirection {
 /// \param direction SortDirection
 /// \return Tensor
 ///
-inline Tensor argsort(const Tensor& x, int axis = -1, SortDirection direction = SortDirection::ASC) {
-  return intrinsic("argsort", x, static_cast<int64_t>(axis), static_cast<int64_t>(direction));
+inline Tensor argsort(const Tensor& x, int axis = -1, SortDirection direction = SortDirection::ASC,
+                      SortStability stability = SortStability::STABLE) {
```
Comment (tzerrell) [inline plaidml/edsl/edsl.h:815]: Based on both TensorFlow's approach and my own intuition, I think we should instead default to the (probably more performant) unstable sort and require the user to explicitly request a stable sort.

**Comment by tzerrell** [comment]
If we do decide to leave the interface as-is and force a single choice between stable & unstable, it's unclear to me whether stable is the correct choice. We're adding it because the OpenVINO TopK tests require a stable sort to pass; however, the _documentation_ for TopK does not mention stability, so it's unclear if this is a requirement or just a (misplaced) assumption of the tests. See OV docs: https://docs.openvinotoolkit.org/latest/openvino_docs_ops_sort_TopK_3.html

**Comment by jbruestle** [comment]
Personally I'd rather only have a stable sort if the perf gap isn't significant.  Some things definitely *require* stable sort, so if we only picked one, that would be the way to go.  I'd also actually tend to make it default if we allow both, since if you get unstable when you were expecting stable, you results will be wrong, whereas if you get stable when you were OK with unstable, all you are doing is losing some perf.  So default stable seems safer to me.  As for the per difference, the perf results here are interesting:

https://stackoverflow.com/questions/810951/how-big-is-the-performance-gap-between-stdsort-and-stdstable-sort-in-practic

Basically, there is a measurable win on the CPU from unstable sort, but it's not huge.  Not sure about other platforms.

**Comment by tzerrell** [comment]
Ok, I'm convinced. Going to stable-sort-only sounds reasonable to me.

**Comment by jbruestle** [comment]
Also, we can add an 'unstable' option later as a perf thing if we want.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 7 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: Using the test case below:
```
TEST_F(CppEdsl, ArgSort1dInt) {
  auto I = Placeholder(DType::INT32, {10});
  auto O = argsort(I, /*axis=*/0);
  auto program = makeProgram("argsort", {I}, {O});
  std::vector<int32_t> input = {
    0, 1, 8, 5, 5, 2, 0, 7, 7, 10,
  }
  std::vector<int32_t> output = {
    0, 6, 1, 5, 3, 4, 7, 8, 2, 9,
  }
  checkExact(program, {input}, {output});
```
Will raise an error:
```
error: 'std.cmpf' op operand #0 must be floating-point-like, but got 'i32'
  note: see current operation: %13 = "std.cmpf"(%11, %12) {predicate = 4 : i64} : (i32, i32) -> i1
```
Integer input tensor is not tested and it seems this will run into the "Assume float" case which causes type mismatch. I guess `elementType.isSignlessInteger()` is missing so I simply follow what `convertAgg` function at line 552 does. In short:
```
if (type.isa<FloatType>()) {
  ...
} else if (type.isSignedInteger()) {
  ...
} else {
  ...
}
```
And the error's gone. Please let me know if there's anything else missing.

**Comment by haoyouab** [comment]
> If we do decide to leave the interface as-is and force a single choice between stable & unstable, it's unclear to me whether stable is the correct choice. We're adding it because the OpenVINO TopK tests require a stable sort to pass; however, the _documentation_ for TopK does not mention stability, so it's unclear if this is a requirement or just a (misplaced) assumption of the tests. See OV docs: https://docs.openvinotoolkit.org/latest/openvino_docs_ops_sort_TopK_3.html

Yes there's no mention about whether `TopK` uses stable sort or not. My assumption is that it's dependent on implementation. In openvino, I run test cases of `TopK` and the original selection sort by Mars failed while the stable one worked correctly so my best guess is OV is implemented in a stable way (selection sort failed anyway).

**Comment by haoyouab** [comment]
I agree that we should keep **ONLY** stable sort in `argsort` for now. I will remove the `stability` flag and make stable the default behavior.

Besides, I found something interesting when I looked into the OV ref implementation of `TopK`. Please check [TopK OV reference implementation](https://github.com/plaidml/openvino/blob/plaidml/ngraph/core/reference/include/ngraph/runtime/reference/topk.hpp#L136) for details. Basically `std::sort` is used, which does not guarantee to be stable. It's OK for the values tensor but when re-arrange indices based on this (see [this](https://github.com/plaidml/openvino/blob/plaidml/ngraph/core/reference/include/ngraph/runtime/reference/topk.hpp#L144)), the indices of two equivalent values could be re-ordered(or not, there's no guarantee). As I mentioned above, in the test cases they are not re-ordered but I think this is dependent on implementations. Perhaps we should change it to `std::stable_sort` to make sure that. Any ideas?

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 7 above)
Comment (tzerrell) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: I think it makes sense for these cases and the cases in line 552 to match. So I think this is a good format for this PR.

If we want to make adjustments, I propose we make a separate PR and change both of these sections (and any other similar sections) in that PR.

**Review by tzerrell — COMMENTED** [review]
IMO this PR is essentially ready for approval, but I'm going to hold off on clicking "Approve" so we can land the sort->stable_sort reference PR and include that repo reference update with in this.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 10]
```diff
@@ -1015,42 +1020,41 @@ struct ArgSortOpConversion : public OpConversionPattern<tile::ArgSortOp> {
...
-        CmpIPredicate dir{};
+      if (elementType.isa<FloatType>()) {
+        CmpFPredicate dir{};
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1024]: Why is there a trailing `{}`?

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 11]
```diff
@@ -1066,16 +1070,27 @@ struct ArgSortOpConversion : public OpConversionPattern<tile::ArgSortOp> {
       // unsorted region, by setting the insertion point after its block.
       rewriter.setInsertionPointAfter(minLoop);
 
-      // Swap the sort position with the minimum position.
       auto finalMinPos =
           rewriter.create<LoadOp>(loc, minIdxVar, zeroIndex).getResult();
       ops[axis] = finalMinPos;
       auto finalMinIdx = rewriter.create<LoadOp>(loc, result, ops).getResult();
+
+      // Move every elements between [sortIV,finalMinPos) a step forward
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1078]: Could rephrase this as: "Move all elements ..." or "Move every element ..."

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 7 above)
Comment (flaub) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: This follows other code that also falls back to unsigned if it's not float or signed. Just FYI, we don't use signless integer types at the edsl/tile level. 

Also, nit: I'd try to break up this function, it's kind of long. Maybe make a helper function that just returns the comparison op (like something that returns the value for `orderPred`).

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 10 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1024]: Actually I didn't add this but just reordered the cases. Anyway it looks unnecessary so I will remove it.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 7 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: The test case above falls back to the last case (presumably unsigned) but before this change, there was a case that explicitly used `else if (type.isUnsignedInteger())` and it didn't match the test. This is why I guess it's a `SignlessInteger` case. If it's not, I'm not sure why it falls back to the last case then.

As for making a helper function, I updated the code and please check.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 11 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Changed to "Move every element ...".

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 12]
```diff
@@ -1105,6 +1085,48 @@ struct ArgSortOpConversion : public OpConversionPattern<tile::ArgSortOp> {
...
+      case tile::SortDirection::asc:
+        pred = CmpFPredicate::OLT;
+        break;
+      case tile::SortDirection::desc:
+        pred = CmpFPredicate::OGT;
+        break;
+      }
+      return rewriter.create<mlir::CmpFOp>(loc, pred, compVal, minVal)
+          .getResult();
+    } else if (type.isSignedInteger()) {
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1104]: no need for `else if` since the first `if` is returning. just `if` would be sufficient

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 7 above)
Comment (flaub) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: I'm curious why we're seeing a signless integer in this case, but it's not a blocker for this PR

Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1018]: I will try to figure it out.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 12 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1104]: OK. Then some other similar code block (like line 552) may need this change, too.

Besides, I'm planning to add the `ArgSort1DInt` test case to this PR.

**Review by tzerrell — APPROVED** [review]
This looks good!
